ghutil/ghutil.go

```diff
@@ -423,94 +433,188 @@ func processPullRequest(ghc *GitHubClient, ctx context.Context, orgName string,
 			}
-			// if PR doesn't have [cla: yes] label, add it.
-			if !hasLabelClaYes {
+		}
+		if issueClaLabelStatus.HasYes {
+			removeLabel(LabelClaYes)
+		}
+		if issueClaLabelStatus.HasNo {
+			removeLabel(LabelClaNo)
+		}
+
+		// No need to add any other CLA-related labels or comments to this PR.
+		return nil
+	} else {
+		if issueClaLabelStatus.HasExternal {
+			logging.Infof("  PR has [%s] label, but shouldn't", LabelClaExternal)
+			removeLabel(LabelClaExternal)
+		} else {
+			logging.Infof("  PR doesn't have [%s] label, and shouldn't", LabelClaExternal)
+			// Nothing to do here.
+		}
+	}
+
+	if pullRequestStatus.Compliant {
+		logging.Info("  PR is CLA-compliant")
+	} else {
+		logging.Info("  PR is NOT CLA-compliant:", pullRequestStatus.NonComplianceReason)
+	}
+
+	// Add or remove [cla: yes] and [cla: no] labels, as appropriate.
+	if pullRequestStatus.Compliant {
+		// if PR has [cla: no] label, remove it.
+		if issueClaLabelStatus.HasNo {
+			removeLabel(LabelClaNo)
+		} else {
+			logging.Infof("  No action needed: [%s] label already missing", LabelClaNo)
+		}
+		// if PR doesn't have [cla: yes] label, add it.
+		if !issueClaLabelStatus.HasYes {
+			if repoClaLabelStatus.HasYes {
 				addLabel(LabelClaYes)
-			} else {
-				logging.Infof("  No action needed: [%s] label already added", LabelClaYes)
 			}
-		} else /* !pullRequestIsCompliant */ {
-			labelsUpdatedForNonCompliance := false
-			// if PR doesn't have [cla: no] label, add it.
-			if !hasLabelClaNo {
+		} else {
+			logging.Infof("  No action needed: [%s] label already added", LabelClaYes)
+		}
+	} else /* !pullRequestIsCompliant */ {
+		shouldAddComment := false
+		// if PR doesn't have [cla: no] label, add it.
+		if !issueClaLabelStatus.HasNo {
+			if repoClaLabelStatus.HasNo {
 				addLabel(LabelClaNo)
-				labelsUpdatedForNonCompliance = true
-			} else {
-				logging.Infof("  No action needed: [%s] label already added", LabelClaNo)
 			}
-			// if PR has [cla: yes] label, remove it.
-			if hasLabelClaYes {
-				removeLabel(LabelClaYes)
-				labelsUpdatedForNonCompliance = true
-			} else {
-				logging.Infof("  No action needed: [%s] label already missing", LabelClaYes)
+			shouldAddComment = true
+		} else {
+			logging.Infof("  No action needed: [%s] label already added", LabelClaNo)
+		}
+		// if PR has [cla: yes] label, remove it.
+		if issueClaLabelStatus.HasYes {
+			removeLabel(LabelClaYes)
+			shouldAddComment = true
+		} else {
+			logging.Infof("  No action needed: [%s] label already missing", LabelClaYes)
+		}
+
+		if shouldAddComment {
+			addComment(pullRequestStatus.NonComplianceReason)
+		}
+	}
+
+	return nil
+}
+
+// IsExternal computes whether the given commit should be processed by this
+// tool, or if it should be covered by an external CLA management tool.
+func IsExternal(commit *github.RepositoryCommit, claSigners config.ClaSigners) bool {
+	if claSigners.External == nil {
+		return false
+	}
+
+	var usernames []string
+	if commit.Author != nil {
```

Comment:
I'll look into making this change, thanks for the suggestion!

Reply:
I tried this suggestion verbatim in `IsExternal()`, but ran into an error:

```
ghutil/ghutil.go:594:26: commit.AuthorLogin undefined (type *github.RepositoryCommit has no field or method AuthorLogin)
```

It looks like I cannot get the type conversion to happen implicitly, so I have to be explicit about it:

```
        if authorLogin := (&Commit{commit}).AuthorLogin(); authorLogin != "" {
                usernames = append(usernames, authorLogin)
        }
```

Even the parentheses are required; it doesn't compile without it. The syntax seems somewhat complex to read here, so I've implemented it slightly differently by creating a standalone method rather than a struct method and added it as a separate commit to this PR.

Please take a look and let me know what you think.

Reply:
For simplicity, I moved this refactoring into a separate PR #50 and will submit it shortly when the tests pass; this makes the current PR much smaller and simpler. Please leave comments on that PR if you have some thoughts on improving that refactoring. Thanks again for the suggestion!